### PR TITLE
Separate exercise & fetch for interfaces from templates

### PR DESCRIPTION
**Pull request by cocreature**
part of #10810

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 1]
```diff
@@ -639,6 +639,17 @@ data Update
     , exeArg        :: !Expr
       -- ^ Argument for the choice.
     }
+  -- | Exercise choice on a contract of an interface given a contract ID.
+  | UExerciseInterface
+    { exeInterface   :: !(Qualified TypeConName)
+      -- ^ Qualified type constructor corresponding to the contract template.
```
Comment (sofiafaro-da) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs:645]:
```suggestion
      -- ^ Qualified type constructor corresponding to the interface.
```

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 2]
```diff
@@ -658,6 +669,14 @@ data Update
       -- ^ Contract id of the contract template instance whose argument shall be
       -- retrieved.
     }
+  -- | Retrieve the argument of an existing contract interface instance.
+  | UFetchInterface
+    { fetInterface   :: !(Qualified TypeConName)
+      -- ^ Qualified type constructor corresponding to the contract template.
```
Comment (sofiafaro-da) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs:675]:
```suggestion
      -- ^ Qualified type constructor corresponding to the interface.
```

**Review by remyhaemmerle-da — APPROVED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 3]
```diff
@@ -773,6 +773,9 @@ private[lf] final class Compiler(
         compileBlock(bindings, body)
       case UpdateFetch(tmplId, coidE) =>
         FetchDefRef(tmplId)(compile(coidE))
+      case UpdateFetchInterface(_, _) =>
+        // TODO https://github.com/digital-asset/daml/issues/10810
+        sys.error("Interfaces not supported")
```
Comment (sofiafaro-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:778]: fine to use FetchDefRef here

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 4]
```diff
@@ -784,6 +787,9 @@ private[lf] final class Compiler(
           choiceId = chId,
           argument = compile(argE),
         )
+      case UpdateExerciseInterface(_, _, _, _) =>
+        // TODO https://github.com/digital-asset/daml/issues/10810
+        sys.error("Interfaces not supported")
```
Comment (sofiafaro-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:792]: fine to use compileExercise (as above), or a ChoiceDefRef directly here

**Review by sofiafaro-da — APPROVED** [review]
Thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 3 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:778]: I think I’d rather wait until that actually works and change it then. At the moment, that won’t do anything sensible.

**Pull request merged**